### DISCUSSION TITLE
Custom Block Entity With A GUI Blinks

### DISCUSSION THREAD

**Discussion by mangodev01** [FabricMC/fabric-api#2477] [Mod Dev Support] [answered]
i wanted to make a diamond crafting table block entity with a gui, but for some reason, when i open this gui, it opens with an expected gui, then it closes, i have a [repository](https://github.com/Gamer069/illia_fabric_1_19_mod) for this

**Comment by warjort** [comment]
Check your run/logs/latest.log there is likely an error in there.

e.g. First thing I find is this is never registered.
https://github.com/Gamer069/illia_fabric_1_19_mod/blob/24b2d32f3e3a51c54d25119f15d79740f593568b/src/main/java/net/illia/illiafabricmod1_19/screen/ModScreenHandlers.java#L9

**Reply by mangodev01** [reply]
so, do i need to remove the 
```
<>
```
?

**Reply by mangodev01** [reply]
> so, do i need to remove the
> 
> ```
> <>
> ```
> 
> ?

or pass in my screen handler type?

**Reply by mangodev01** [reply]
in the logs, there is no errors, warnings, etc.

**Reply by warjort** [reply]
Removing that syntax doesn't register that ScreenHandlerType?
You can see how vanilla does it in ScreenHandlerType.register()

**Reply by mangodev01** [reply]
thanks :)

**Reply by mangodev01** [reply]
but where is the ScreenHandlerType.register()?

**Reply by warjort** [reply]
`net.minecraft.screen.ScreenHandlerType`

**Reply by mangodev01** [reply]
thanks

**Reply by warjort** [reply]
Have you run `gradlew gensources` so the minecraft source is available in your ide?

**Reply by mangodev01** [reply]
of course

**Reply by mangodev01** [reply]
nvm, i found it

**Reply by mangodev01** [reply]
is it possible to set the ScreenHandlerType in the screen handler to "CRAFTING" so i will not need to implement like 1 million methods, will this work or will the registration fail?

**Reply by warjort** [reply]
The ScreenHandlerType identifies the screen to use. 

It is possible to reuse the CraftingScreenHandler by subclassing it.
You will need to override;

* getType() to return your ScreenHandlerType so it knows to use a different screen
* canUse(player) to check for your Block instead of Blocks.CRAFTING_TABLE

I assume that your screen will be like the vanilla crafting table screen
Otherwise the answer is no.

**Reply by mangodev01** [reply]
thanks

**Reply by mangodev01** [reply]
but how do i fix the blinking?

**Reply by warjort** [reply]
You haven't updated your repository with the latest changes.
Or if you have, you haven't fixed the registration problem.

**Reply by mangodev01** [reply]
thanks, but how do i register it properly?
```
Registry.register(Registry.SCREEN_HANDLER, new Identifier(IlliaFabricMod1_19.MOD_ID, "diamond_crafting_table_screen_handler"), DIAMOND_CRAFTING_TABLE_SCREEN_HANDLER);
```
like this?

**Reply by mangodev01** [reply]
or is this just for the 
```
ScreenHandlerType.register();
```
to function properly?

**Reply by warjort** [reply]
Looks ok to me.

**Comment by warjort** [comment, marked as answer]
Your next problem is here:
https://github.com/Gamer069/illia_fabric_1_19_mod/blob/1d1864562ece267a6bbf5f35f2f9c0d9cfd1cdd6/src/main/java/net/illia/illiafabricmod1_19/block/custom/DiamondCraftingTable.java#L77

You are telling it to display the vanilla crafting table screen.
This will check your block is a vanilla crafting table and close the screen if it is not.
See CraftingScreenHandler.canUse()

You want it to use your ScreenHandler
You are using a BlockEntity - the vanilla crafting table is not a block entity so the code is different.

You want something like the following in DiamondCraftingTable (untested code):
````
	@Override
	public ActionResult onUse(BlockState state, World world, BlockPos pos, PlayerEntity player, Hand hand, BlockHitResult hit) {
		if (!world.isClient) {
	            BlockEntity blockEntity = world.getBlockEntity(pos);
	            if (blockEntity instanceof DiamondCraftingTableBlockEntity diamondCraftingTableBlockEntity) {
	                player.openHandledScreen(diamondCraftingTableBlockEntity);
	            }
		}
		return ActionResult.SUCCESS;
	}
````

You don't need that createScreenHandlerFactory() method in that class. Your BlockEntity is a NamedScreenHandlerFactory
https://github.com/Gamer069/illia_fabric_1_19_mod/blob/1d1864562ece267a6bbf5f35f2f9c0d9cfd1cdd6/src/main/java/net/illia/illiafabricmod1_19/block/entity/DiamondCraftingTableBlockEntity.java#L24

**Reply by mangodev01** [reply]
btw, if the crafting table isn't a block entity, then what is it?, thanks

**Reply by warjort** [reply]
It's just a normal block. 
It has no internal inventory or anything that needs saving.

**Reply by mangodev01** [reply]
> inventory

thanks, also why does the inventory not render properly?
(when hovered over a slot, for some reason there is some offset, there are "invisible rendered slots"), i can send a screenshot

**Reply by mangodev01** [reply]
here are the screenshots:
https://user-images.githubusercontent.com/106342228/186422488-fc4a7098-9691-406b-b1fc-46aec4d13374.png
https://user-images.githubusercontent.com/106342228/186422690-2a73fe5f-ae48-4a73-8d8b-f6807dcb379b.png

**Reply by warjort** [reply]
https://github.com/Gamer069/illia_fabric_1_19_mod/blob/1e6559bb6b05ba20041d676d874179d0ca829509/src/main/java/net/illia/illiafabricmod1_19/screen/DiamondCraftingTableScreenHandler.java#L80

You have a 9x9 player inventory when it should be 3x9 :-)

**Reply by mangodev01** [reply]
so, do i change 9 to 3 everywhere?

**Reply by warjort** [reply]
No. Look at other vanilla screen handlers.
A player's inventory is 27 slots, usually organised as 9 across and 3 down.

You need to understand what this is doing. 
If you get a bug report for your mod you need to be able to fix it yourself.
You can't just get us to fix all your bugs.

**Reply by mangodev01** [reply]
thanks, i will look at vanilla screen handlers now

**Reply by mangodev01** [reply]
thanks, it worked now

**Reply by mangodev01** [reply]
also, how do i make my recipe type work?

**Reply by warjort** [reply]
Can you start a new discussion topic.
This one is already answered and it is way too long.

**Reply by mangodev01** [reply]
ok, i made it [here](https://github.com/orgs/FabricMC/discussions/2484)
